crates/lib3h/src/engine/ghost_engine.rs

```diff
@@ -12,3 +37,1064 @@ pub type GhostEngineParentWrapper<Core, Engine, EngineError> = GhostParentWrappe
+            ClientToLib3h,
+            ClientToLib3hResponse,
+            Lib3hToClient,
+            Lib3hToClientResponse,
+            Lib3hError,
+        >,
+    > {
+        std::mem::replace(&mut self.client_endpoint, None)
+    }
+    // END BOILER PLATE--------------------------
+
+    fn process_concrete(&mut self) -> GhostResult<WorkWasDone> {
+        // START BOILER PLATE--------------------------
+        // always run the endpoint process loop
+        detach_run!(&mut self.lib3h_endpoint, |cs| { cs.process(self) })?;
+        // END BOILER PLATE--------------------------
+
+        // process any messages from the client to us
+        for msg in self.lib3h_endpoint.as_mut().drain_messages() {
+            self.handle_msg_from_client(msg)?;
+        }
+
+        /*
+                let outbox: Vec<Lib3hServerProtocol> = Vec::new();
+                let (net_did_work, mut net_outbox) = self.process_multiplexer()?;
+                outbox.append(&mut net_outbox);
+                // Process the space layer
+                let mut p2p_output = self.process_space_gateways()?;
+                outbox.append(&mut p2p_output);
+                // Hack
+                let (ugly_did_work, mut ugly_outbox) = self.process_ugly();
+                outbox.append(&mut ugly_outbox);
+        */
+
+        Ok(true.into())
+    }
+}
+
+/// Drop
+impl<'engine> Drop for GhostEngine<'engine> {
+    fn drop(&mut self) {
+        self.shutdown().unwrap_or_else(|e| {
+            warn!("Graceful shutdown failed: {}", e);
+        });
+    }
+}
+
+/// Private
+impl<'engine> GhostEngine<'engine> {
+    /// Called on drop.
+    /// Close all connections gracefully
+    fn shutdown(&mut self) -> Lib3hResult<()> {
+        let /*mut*/ result: Lib3hResult<()> = Ok(());
+
+        // FIXME
+        //        for space_gatway in self.space_gateway_map.values_mut() {
+        //            let res = space_gatway.as_transport_mut().close_all();
+        //            // Continue closing connections even if some failed
+        //            if let Err(e) = res {
+        //                if result.is_ok() {
+        //                    result = Err(e.into());
+        //                }
+        //            }
+        //        }
+        //        self.multiplexer
+        //            .as_transport_mut()
+        //            .close_all()
+        //            .map_err(|e| {
+        //                error!("Closing of some connection failed: {:?}", e);
+        //                e
+        //            })?;
+
+        result
+    }
+
+    /// Process connect events by sending them to the multiplexer
+    fn handle_connect(&mut self, data: ConnectData) -> GhostResult<()> {
+        let cmd =
+            GatewayRequestToChild::Transport(transport::protocol::RequestToChild::SendMessage {
+                uri: data.peer_uri,
+                payload: Opaque::new(),
+            });
+        self.multiplexer.publish(Lib3hSpan::todo(), cmd)
```

Comment:
I think this needs to be a request so we can send back the connect response on the lib3h protocol.

Reply:
adding this a follow up because we have to integrate this with the bootstrap phase requests too.